Allow teleport destinations as restart points for start zones

braem opened the pull request:
Closes #313 

- Does a slight refactoring of the start zone trigger code
- Adds a remote destination to the start zone trigger, much like teleport triggers.
  - destination is teleported to if there is no start mark. (start mark > tele dest > middle of zone teleport)
- Updates zonmaker with the change

Opted not to add the base zone trigger as it would show up for end & checkpoint zones.
Would it be a good idea to add this to stage zones as well?

### Checklist
- [x] **I have thoroughly tested all of the code I have modified/added/removed to ensure something else did not break**
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] If I have added or modified any visual assets (models, materials, panels, effects, etc), I have taken screenshots / videos of them and attached them to this PR directly (screenshots uploaded through github, videos uploaded to youtube and linked)
- [x] If I have modified any console command, console variable, or momentum entity, I have opened an issue (or a PR) for it in the [Momentum Mod documentation repository](https://github.com/momentum-mod/docs)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

SCell555 commented on mp/src/game/server/momentum/mom_triggers.h:
```diff
@@ -306,6 +306,8 @@ class CTriggerTimerStart : public CTriggerZone
     int GetLimitSpeedType() const { return m_iLimitSpeedType; }
     void SetLimitSpeedType(const int type) { m_iLimitSpeedType = type; }
+    const char *GetTeleDestName() const { return m_strTeleDestName.ToCStr(); }
+    void SetTeleDestName(const char *pszName) { m_strTeleDestName = MAKE_STRING(pszName); }
```
This should use `AllocPooledString`, don't hold temporary string pointers

SCell555 commented on mp/src/game/server/momentum/mom_triggers.h:
```diff
@@ -320,6 +322,9 @@ class CTriggerTimerStart : public CTriggerZone
     int m_iLimitSpeedType;
+
+    string_t m_strTeleDestName;
+    CBaseEntity *m_pTeleDest;
```
EHANDLE would be better, but it probably doesn't matter. Nobody is going to tamper with entity after it's found.

Gocnak commented on mp/src/game/server/momentum/mom_triggers.h:
```diff
@@ -320,6 +322,9 @@ class CTriggerTimerStart : public CTriggerZone
     int m_iLimitSpeedType;
+
+    string_t m_strTeleDestName;
+    EHANDLE m_pTeleDest;
```
```suggestion
    EHANDLE m_hTeleDest;
```

xen-000 commented:
> Would it be a good idea to add this to stage zones as well?

Yes, any zone that can explicitly be teleported to should have this.

braem commented:
- Moved logic to `CTriggerZone`, only exposing to start & stage zones.
- In KV, moved out of `"zoneProps"` -> `"properties"`
- Changed `Teleport` to `ManualTeleport` in stage restarting, making use of the added `bStopTimer` parameter.

Gocnak commented on mp/src/game/server/momentum/mom_triggers.h:
```diff
@@ -215,14 +214,19 @@ class CBaseMomZoneTrigger : public CBaseMomentumTrigger
-    int GetZoneNumber() const { return m_iZoneNumber; };
+    int GetZoneNumber() const { return m_iZoneNumber; }
+
+    virtual bool IsTeleportableTo() { return false; }
```
```suggestion
    bool IsTeleportableTo() const { return GetType() == ZONE_TYPE_START || GetType() == ZONE_TYPE_STAGE; }
```

braem replied:
Had to make `GetZoneType()` `const` to do this but it should've already been `const` anyhow.

Gocnak commented on mp/src/game/server/momentum/mom_player.cpp:
```diff
@@ -1879,9 +1879,17 @@ void CMomentumPlayer::TimerCommand_Restart(int track)
-            // Don't set angles if still in start zone.
-            QAngle ang = pStart->GetLookAngles();
-            ManualTeleport(&pStart->GetRestartPosition(), (pStart->HasLookAngles() ? &ang : nullptr), &vec3_origin);
+            const auto pZoneTeleDest = pStart->GetTeleDest();
+            if (pZoneTeleDest)
+            {
+                ManualTeleport(&pZoneTeleDest->GetAbsOrigin(), &pZoneTeleDest->GetAbsAngles(), &vec3_origin);
+            }
+            else
+            {
+                // Don't set angles if still in start zone.
+                QAngle ang = pStart->GetLookAngles();
+                ManualTeleport(&pStart->GetRestartPosition(), pStart->HasLookAngles() ? &ang : nullptr, &vec3_origin);
+            }
```
The only thing changing here is what's passed in for target dest and angles. Could be good to instead make the logic about selecting the position and angles instead of calling ManualTeleport multiple times:

```suggestion
            Vector *pTargetDest = nullptr;
            QAngle *pTargetAng = nullptr;
            const auto pZoneTeleDest = pStart->GetTeleDest();
            if (pZoneTeleDest)
            {
                pTargetDest = pZoneTeleDest->GetAbsOrigin();
                pTargetAng = pZoneTeleDest->GetAbsAngles();
            }
            else
            {
                pTargetDest = &pStart->GetRestartPosition();
                pTargetAng = pStart->HasLookAngles() ? &pStart->GetLookAngles() : nullptr;
            }

            ManualTeleport(pTargetDest, pTargetAng, &vec3_origin);
```

braem replied:
Did this for restart stage as well.

Gocnak commented on mp/src/game/server/momentum/mom_player.cpp:
```diff
@@ -1879,9 +1879,12 @@ void CMomentumPlayer::TimerCommand_Restart(int track)
+            const auto pZoneTeleDest = pStart->GetTeleDest();
+
+            const Vector *pTargetDest = pZoneTeleDest ? &pZoneTeleDest->GetAbsOrigin() : &pStart->GetRestartPosition();
+            const QAngle *pTargetAng = pZoneTeleDest ? &pZoneTeleDest->GetAbsAngles() : pStart->HasLookAngles() ? &pStart->GetLookAngles() : nullptr;
```
I'm gonna approve but next time use if/else for a ternery this large
